From d61f3f6f5a4b438ea34c22e978219cad590fabd3 Mon Sep 17 00:00:00 2001
From: Michal Hrusecky <michal.hrusecky@turris.com>
Date: Thu, 20 Mar 2025 10:50:09 +0100
Subject: [PATCH] watchcat: Add option to restart logical interface

Some interfaces are harder to restart then just link down/up. For
example VPNs or NCM modems.
---
 utils/watchcat/files/watchcat.init |  3 ++-
 utils/watchcat/files/watchcat.sh   | 22 +++++++++++++++++-----
 2 files changed, 19 insertions(+), 6 deletions(-)

diff --git a/utils/watchcat/files/watchcat.init b/utils/watchcat/files/watchcat.init
index dcaa8ece5..6af22db1f 100644
--- a/utils/watchcat/files/watchcat.init
+++ b/utils/watchcat/files/watchcat.init
@@ -39,6 +39,7 @@ config_watchcat() {
 	config_get forcedelay "$1" forcedelay "60"
 	config_get pingsize "$1" pingsize "standard"
 	config_get interface "$1" interface
+	config_get logical_interface "$1" logical_interface
 	config_get mmifacename "$1" mmifacename "null"
 	config_get_bool unlockbands "$1" unlockbands "0"
 	config_get addressfamily "$1" addressfamily "any"
@@ -110,7 +111,7 @@ config_watchcat() {
 		;;
 	restart_iface)
 		procd_open_instance "watchcat_${1}"
-		procd_set_param command /usr/bin/watchcat.sh "restart_iface" "$period" "$pinghosts" "$pingperiod" "$pingsize" "$interface" "$mmifacename" "$unlockbands" "$addressfamily"
+		procd_set_param command /usr/bin/watchcat.sh "restart_iface" "$period" "$pinghosts" "$pingperiod" "$pingsize" "$interface" "$mmifacename" "$unlockbands" "$logical_interface" "$addressfamily"
 		procd_set_param respawn "${respawn_threshold:-3600}" "${respawn_timeout:-5}" "${respawn_retry:-5}"
 		procd_close_instance
 		;;
diff --git a/utils/watchcat/files/watchcat.sh b/utils/watchcat/files/watchcat.sh
index 0f300a4a4..0735145c9 100644
--- a/utils/watchcat/files/watchcat.sh
+++ b/utils/watchcat/files/watchcat.sh
@@ -87,6 +87,13 @@ watchcat_restart_network_iface() {
 	ip link set "$1" up
 }
 
+watchcat_restart_logical_network_iface() {
+	logger -p daemon.info -t "watchcat[$$]" "Restarting logical network interface: \"$1\"."
+	ifdown "$1"
+	ifup "$1"
+}
+
+
 watchcat_run_script() {
 	logger -p daemon.info -t "watchcat[$$]" "Running script \"$1\" for network interface: \"$2\"."
 	"$1" "$2"
@@ -105,8 +112,9 @@ watchcat_monitor_network() {
 	iface="$5"
 	mm_iface_name="$6"
 	mm_iface_unlock_bands="$7"
-	address_family="$8"
-	script="$9"
+	logical_interface="$8"
+	address_family="$9"
+	script="${10}"
 
 	time_now="$(cat /proc/uptime)"
 	time_now="${time_now%%.*}"
@@ -152,6 +160,8 @@ watchcat_monitor_network() {
 			else
 				if [ "$script" != "" ]; then
 					logger -p daemon.info -t "watchcat[$$]" "Could not reach $host via \"$iface\" for \"$((time_now - time_lastcheck_withinternet))\" seconds. Running script after reaching \"$failure_period\" seconds"
+				elif [ "$logical_iface" != "" ]; then
+					logger -p daemon.info -t "watchcat[$$]" "Could not reach $host for \"$((time_now - time_lastcheck_withinternet))\" seconds. Restarting \"$logical_iface\" after reaching \"$failure_period\" seconds"
 				elif [ "$iface" != "" ]; then
 					logger -p daemon.info -t "watchcat[$$]" "Could not reach $host via \"$iface\" for \"$((time_now - time_lastcheck_withinternet))\" seconds. Restarting \"$iface\" after reaching \"$failure_period\" seconds"
 				else
@@ -167,7 +177,9 @@ watchcat_monitor_network() {
 				if [ "$mm_iface_name" != "" ]; then
 					watchcat_restart_modemmanager_iface "$mm_iface_name" "$mm_iface_unlock_bands"
 				fi
-				if [ "$iface" != "" ]; then
+				if [ "$logical_iface" != "" ]; then
+					watchcat_restart_logical_network_iface "$logical_iface"
+				elif [ "$iface" != "" ]; then
 					watchcat_restart_network_iface "$iface"
 				else
 					watchcat_restart_all_network
@@ -257,11 +269,11 @@ ping_reboot)
 	;;
 restart_iface)
 	# args from init script: period pinghosts pingperiod pingsize interface mmifacename unlockbands addressfamily
-	watchcat_monitor_network "$2" "$3" "$4" "$5" "$6" "$7" "$8" "$9" ""
+	watchcat_monitor_network "$2" "$3" "$4" "$5" "$6" "$7" "$8" "$9" "${10}" ""
 	;;
 run_script)
 	# args from init script: period pinghosts pingperiod pingsize interface addressfamily script
-	watchcat_monitor_network "$2" "$3" "$4" "$5" "$6" "" "" "$7" "$8"
+	watchcat_monitor_network "$2" "$3" "$4" "$5" "$6" "" "" "" "$7" "$8"
 	;;
 *)
 	echo "Error: invalid mode selected: $mode"
-- 
2.52.0

